node/core/approval-voting/src/approval_checking.rs

```diff
@@ -0,0 +1,619 @@
+	// no-shows encountered along the way. Once all of the no-shows we were previously aware
+	// of are covered, we then progress to cover the no-shows we encountered while covering those,
+	// and so on.
+	#[derive(Debug)]
+	enum State {
+		// (assignments, no-shows)
+		InitialCount(usize, usize),
+		// (assignments, covered no-shows, covering no-shows, uncovered no-shows),
+		CoverNoShows(usize, usize, usize, usize),
+	}
+
+	impl State {
+		fn output(
+			&self,
+			tranche: DelayTranche,
+			tranche_now: DelayTranche,
+			needed_approvals: usize,
+			n_validators: usize,
+		) -> RequiredTranches {
+			match *self {
+				State::InitialCount(assignments, no_shows) =>
+					if assignments >= needed_approvals && no_shows == 0 {
+						RequiredTranches::Exact(tranche, 0)
+					} else {
+						// This happens only if there are not enough assignments, period.
+						//
+						// However, within the context of this function in particular,
+						// it can happen if there are enough assignments but there are no-shows.
+						// The calling code would have already transitioned the state to
+						// `CoverNoShows` before invoking this function, so that is not possible
+						// in practice.
+						//
+						// In the case of not-enough assignments, we want assignments to broadcast based
+						// on timer, so we treat it as though there are no uncovered no-shows.
+						if no_shows == 0 {
+							RequiredTranches::Pending(tranche_now)
+						} else if assignments < needed_approvals {
+							// This branch is only hit if there are no-shows before there
+							// are enough initial assignments.
+							RequiredTranches::Pending(tranche_now)
+						} else {
+							// This branch is never hit, as explained above.
+							RequiredTranches::Pending(tranche + no_shows as DelayTranche)
+						}
+					},
+				State::CoverNoShows(total_assignments, covered, covering, uncovered) =>
+					if covering == 0 && uncovered == 0 {
+						RequiredTranches::Exact(tranche, covered)
+					} else if total_assignments + covering + uncovered >= n_validators  {
+						RequiredTranches::All
+					} else {
+						RequiredTranches::Pending(tranche + (covering + uncovered) as DelayTranche)
+					},
+			}
+		}
+	}
+
+	let tick_now = tranche_now as Tick + block_tick;
+	let n_validators = approval_entry.n_validators();
+
+	approval_entry.tranches().iter()
+		.take_while(|t| t.tranche() <= tranche_now)
+		.scan(Some(State::InitialCount(0, 0)), |state, tranche| {
+			// The `Option` here is used for early exit.
+			let s = match state.take() {
+				None => return None,
+				Some(s) => s,
+			};
+
+			let n_assignments = tranche.assignments().len();
+
+			// count no-shows. An assignment is a no-show if there is no corresponding approval vote
+			// after a fixed duration.
+			let no_shows = tranche.assignments().iter().filter(|(v_index, tick)| {
+				tick + no_show_duration <= tick_now
+					&& approvals.get(*v_index as usize).map(|b| !*b).unwrap_or(true)
+			}).count();
```

Comment:
I'd one no show delay value that increased during when I implemented this iteration over tranches.  It served two purposes, delayed triggering our own assignment announcements, and terminate this loop early.  

In this version, I'm suspicious that `tick_now` should be reduced by `no_show_duration` every time we take another batch of no-shows.